File: algorithms/73. Set Matrix Zeroes/solution.cc
class Solution {
public:
    void setZeroes(vector<vector<int>>& matrix) {
        int m = matrix.size(), n = m ? matrix[0].size() : 0, col0 = 1;
        for ( int i = 0; i < m; ++i ) {
            if ( 0 == matrix[i][0] ) {
                col0 = 0;
            }
            for ( int j = 1; j < n; ++j ) {
                if ( 0 == matrix[i][j] ) {
                    matrix[i][0] = matrix[0][j] = 0;
                }
            }
        }
        for ( int i = m - 1; i >= 0; --i ) {
            for ( int j = 1; j < n; ++j ) {
                if ( 0 == matrix[i][0] || 0 == matrix[0][j] ) {
                    matrix[i][j] = 0;
                }
            }
            if ( 0 == col0 ) {
                matrix[i][0] = 0;
            }
        }
    }
};

File: algorithms/73. Set Matrix Zeroes/solution.py
class Solution(object):
    def setZeroes(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: void Do not return anything, modify matrix in-place instead.
        """
        m, n = len(matrix), len(matrix[0]) if matrix else 0
        col0 = 1
        for i in xrange(m):
            if matrix[i][0] == 0:
                col0 = 0
            for j in xrange(1, n):
                if matrix[i][j] == 0:
                    matrix[i][0] = matrix[0][j] = 0
        for i in xrange(m-1, -1, -1):
            for j in xrange(1, n):
                if matrix[0][j] == 0 or matrix[i][0] == 0:
                    matrix[i][j] = 0
            if col0 == 0:
                matrix[i][0] = 0
